### PR TITLE
feat(core): add fragmentation filter

### LINKED ISSUE #353: How to execute tests against only 1 shard of X shards?
How can I have marathon split up tests in a deterministic way? For example, let’s say a module has 100 tests, I want to split them into 4 different shards, and run tests against one of those shards (so only 25 of the tests)?

previously I did this with Spoon using `--numShards 4 --shard-index 0`

### PR DESCRIPTION
**Pull request by Malinskiy**
This is a test filter similar to sharded test execution that AOSP provides https://source.android.com/devices/tech/test_infra/tradefed/architecture/advanced/sharding

It is intended to be used in situations where it is not possible to connect multiple execution devices to marathon, e.g. CI setup that can schedule parallel jobs each containing a single execution device.

This is a dynamic programming technique, hence the results will be sub-optimal compared to connecting multiple devices to the same test run

Apply as following:
```yaml
filteringConfiguration:
  allowlist:
    - type: "fragmentation"
      index: 0
      count: 10
```

If you want to dynamically pass the index of the test run you can use yaml envvar interpolation, e.g.:
```yaml
filteringConfiguration:
  allowlist:
    - type: "fragmentation"
      index: ${MARATHON_FRAGMENT_INDEX}
      count: ${MARATHON_FRAGMENT_COUNT}
```

and then execute the testing as following:

```bash
$ MARATHON_FRAGMENT_COUNT=10 MARATHON_FRAGMENT_INDEX=0 marathon
```

The implementation currently makes a stable decision based on the FQTN. Assuming the count of fragments doesn't change, the test will always be executed in the fragment under the same index. Essentially this makes newly added tests randomly distributed between fragments while guaranteeing that tests are tied to a specific fragment. This execution model doesn't account however for the test duration so an expected edge case is that a particular fragment might exist that will be executed for much longer than other fragments.

Fixes #353

### CODE REVIEW

File: docs/_posts/2018-11-19-configuration.md [hunk 1]
```diff
@@ -120,7 +120,8 @@ Here you will find a list of currently supported configuration parameters and ex
 additional parameters might not be supported by all vendor modules. If you find that something doesn't work - please submit an issue for a
 vendor module at fault.
 
-* TOC {:toc}
+* TOC
```
Comment (Malinskiy) [inline docs/_posts/2018-11-19-configuration.md:123]: Autoformatting removed TOC

File: core/src/main/kotlin/com/malinskiy/marathon/extension/StringExtensions.kt [hunk 2]
```diff
@@ -1,9 +1,20 @@
...
     } else {
         "$prefix.$this"
     }
 }
+
+fun String.md5(): BigInteger {
+    return BigInteger(MD5.md5.digest(toByteArray(charset = Charsets.UTF_8)))
+}
+
+object MD5 {
```
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/extension/StringExtensions.kt:18]: Not sure if we want to separate this yet into a different class

File: core/src/main/kotlin/com/malinskiy/marathon/execution/filter/FragmentationFilter.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        actualRemainder == remainder
+    }
+
+    override fun validate() {
+        if (index < 0) throw ConfigurationException("Fragment index [$index] should be >= 0")
+        if (count < 0) throw ConfigurationException("Fragment count [$count] should be >= 0")
+        if (index >= count) throw ConfigurationException("Fragment index [$index] should be less than count [$count]")
+
+        logger.info { "Executing test fragment $index out of $count" }
+        logger.warn {
```
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/execution/filter/FragmentationFilter.kt:46]: This warning should encourage people to migrate to different setups

**Review by Vacxe — APPROVED** [review]
